File: lib/topic-guard.ts
import "server-only";

import { chatCompletion } from "./openai-client";
import { openaiConfigured } from "./openai-config";

export type TopicScope = "on_topic" | "off_topic" | "unclear";

const ON_TOPIC_PATTERNS: RegExp[] = [
  /\bpbo\b/i,
  /\bngo\b/i,
  /\bcso\b/i,
  /\bcreco\b/i,
  /\bicnl\b/i,
  /public benefit/i,
  /non[- ]?profit/i,
  /nonprofit/i,
  /registration/i,
  /register/i,
  /sajili/i,
  /mahitaji/i,
  /shirika/i,
  /faida ya umma/i,
  /pbo act/i,
  /organizations act/i,
  /organisations act/i,
  /regulatory authority/i,
  /compliance/i,
  /governance/i,
  /trustee/i,
  /board member/i,
  /civic space/i,
  /civil society/i,
  /kenya.*(act|law|legal)/i,
  /(act|law).{0,20}kenya/i,
  /tax exempt/i,
  /fundraising/i,
  /audit/i,
  /annual return/i,
  /by-?laws/i,
  /constitution/i,
  /memorandum/i,
  /charitable/i,
  /community organis/i,
  /community organiz/i,
];

const OFF_TOPIC_PATTERNS: RegExp[] = [
  /\bweather\b/i,
  /\bfootball\b/i,
  /\bsoccer\b/i,
  /\brecipe\b/i,
  /\bbitcoin\b/i,
  /\bcrypto(currency)?\b/i,
  /\bstock market\b/i,
  /\bwrite (me )?(a )?(poem|story|essay|code)\b/i,
  /\bpython\b/i,
  /\bjavascript\b/i,
  /\bhomework\b/i,
  /\bcapital of\b/i,
  /\bwho (is|was) (the )?(president|king|queen) of (?!kenya)/i,
  /\bdating\b/i,
  /\bmovie\b/i,
  /\bnetflix\b/i,
];

export const OFF_TOPIC_REFUSAL =
  "This tool only answers questions about **Kenyan Public Benefit Organizations**, the **PBO Act**, **NGO/CSO registration and compliance**, and **CRECO-related civic guidance**. Please rephrase your question in that context, browse **Topics**, or contact **CRECO Kenya**.";

export function heuristicTopicScope(question: string, wikiMatchScore: number): TopicScope {
  const q = question.trim();
  if (q.length < 3) return "off_topic";

  if (OFF_TOPIC_PATTERNS.some((re) => re.test(q))) {
    return "off_topic";
  }

  if (wikiMatchScore >= 2) {
    return "on_topic";
  }

  const onHits = ON_TOPIC_PATTERNS.filter((re) => re.test(q)).length;
  if (onHits >= 1) {
    return "on_topic";
  }

  const terms = q.toLowerCase().match(/\w{4,}/g) ?? [];
  const genericNgOHints = ["organization", "organisation", "charity", "foundation", "association"];
  if (terms.some((t) => genericNgOHints.some((h) => t.includes(h)))) {
    return "unclear";
  }

  return "unclear";
}

async function classifyWithModel(question: string): Promise<TopicScope> {
  const reply = await chatCompletion(
    [
      {
        role: "system",
        content: `You gate questions for CRECO Kenya's PBO Act guidance site.
ON TOPIC: Kenya PBO Act 2013, PBO/NGO/CSO registration or compliance, CRECO, ICNL, civic space, nonprofit governance in Kenya.
OFF TOPIC: all other subjects (other countries, general trivia, tech help, entertainment, unrelated personal advice).
Reply with exactly one word: ON or OFF.`,
      },
      { role: "user", content: question.slice(0, 500) },
    ],
    0,
    8,
  );

  if (!reply) return "unclear";
  const normalized = reply.trim().toUpperCase();
  if (normalized.startsWith("ON")) return "on_topic";
  if (normalized.startsWith("OFF")) return "off_topic";
  return "unclear";
}

/** Decide if we may use AI or supplemental answers (saves tokens on off-topic questions). */
export async function resolveTopicScope(
  question: string,
  wikiMatchScore: number,
): Promise<TopicScope> {
  const heuristic = heuristicTopicScope(question, wikiMatchScore);
  if (heuristic === "on_topic" || heuristic === "off_topic") {
    return heuristic;
  }

  if (!openaiConfigured()) {
    return wikiMatchScore > 0 ? "on_topic" : "off_topic";
  }

  const modelScope = await classifyWithModel(question);
  if (modelScope !== "unclear") return modelScope;

  return wikiMatchScore > 0 ? "on_topic" : "off_topic";
}

File: prototype/backend/app/topic_guard.py
"""Scope gate — refuse off-topic questions before calling OpenAI."""

from __future__ import annotations

import re

OFF_TOPIC_REFUSAL = (
    "This tool only answers questions about **Kenyan Public Benefit Organizations**, "
    "the **PBO Act**, **NGO/CSO registration and compliance**, and **CRECO-related civic guidance**. "
    "Please rephrase your question in that context, browse **Topics**, or contact **CRECO Kenya**."
)

ON_TOPIC_PATTERNS = [
    re.compile(p, re.I)
    for p in [
        r"\bpbo\b",
        r"\bngo\b",
        r"\bcso\b",
        r"\bcreco\b",
        r"\bicnl\b",
        r"public benefit",
        r"non[- ]?profit",
        r"registration",
        r"register",
        r"sajili",
        r"pbo act",
        r"compliance",
        r"civic space",
        r"civil society",
        r"tax exempt",
        r"by-?laws",
        r"charitable",
    ]
]

OFF_TOPIC_PATTERNS = [
    re.compile(p, re.I)
    for p in [
        r"\bweather\b",
        r"\bfootball\b",
        r"\brecipe\b",
        r"\bbitcoin\b",
        r"\bwrite (me )?(a )?(poem|story|code)\b",
        r"\bpython\b",
        r"\bcapital of\b",
        r"\bmovie\b",
    ]
]


def heuristic_topic_scope(question: str, wiki_match_score: int) -> str:
    q = question.strip()
    if len(q) < 3:
        return "off_topic"
    if any(p.search(q) for p in OFF_TOPIC_PATTERNS):
        return "off_topic"
    if wiki_match_score >= 2:
        return "on_topic"
    if any(p.search(q) for p in ON_TOPIC_PATTERNS):
        return "on_topic"
    return "unclear"


def classify_with_model(question: str, client, model: str) -> str:
    try:
        response = client.chat.completions.create(
            model=model,
            temperature=0,
            max_tokens=8,
            messages=[
                {
                    "role": "system",
                    "content": (
                        "You gate questions for CRECO Kenya's PBO Act guidance site. "
                        "ON TOPIC: Kenya PBO Act, PBO/NGO/CSO registration or compliance, CRECO, civic space. "
                        "OFF TOPIC: everything else. Reply with exactly one word: ON or OFF."
                    ),
                },
                {"role": "user", "content": question[:500]},
            ],
        )
        text = (response.choices[0].message.content or "").strip().upper()
        if text.startswith("ON"):
            return "on_topic"
        if text.startswith("OFF"):
            return "off_topic"
    except Exception:
        pass
    return "unclear"


def resolve_topic_scope(question: str, wiki_match_score: int, *, openai_ok: bool, client=None, model: str = "") -> str:
    heuristic = heuristic_topic_scope(question, wiki_match_score)
    if heuristic in ("on_topic", "off_topic"):
        return heuristic
    if not openai_ok or client is None:
        return "on_topic" if wiki_match_score > 0 else "off_topic"
    model_scope = classify_with_model(question, client, model)
    if model_scope != "unclear":
        return model_scope
    return "on_topic" if wiki_match_score > 0 else "off_topic"
